File: packages/sdk/src/utils/webhook.ts
/**
 * Webhook signature verification utilities
 */

import * as crypto from 'crypto';
import { WebhookVerificationError } from '../errors';
import type { WebhookEvent } from '../types';

/**
 * Verifies webhook signature and returns parsed event
 * @param payload Raw request body as string or buffer
 * @param signature Signature from z402-signature header
 * @param secret Your webhook secret
 * @returns Parsed webhook event
 * @throws WebhookVerificationError if signature is invalid
 */
export function verifyWebhook(
  payload: string | Buffer,
  signature: string,
  secret: string
): WebhookEvent {
  if (!signature) {
    throw new WebhookVerificationError('Missing signature header');
  }

  if (!secret) {
    throw new WebhookVerificationError('Webhook secret not provided');
  }

  // Parse signature (format: t=timestamp,v1=signature)
  const parts = signature.split(',');
  const timestamp = parts.find((p) => p.startsWith('t='))?.split('=')[1];
  const providedSignature = parts.find((p) => p.startsWith('v1='))?.split('=')[1];

  if (!timestamp || !providedSignature) {
    throw new WebhookVerificationError('Invalid signature format');
  }

  // Check timestamp is within 5 minutes
  const timestampMs = parseInt(timestamp, 10) * 1000;
  const now = Date.now();
  const tolerance = 5 * 60 * 1000; // 5 minutes

  if (Math.abs(now - timestampMs) > tolerance) {
    throw new WebhookVerificationError('Signature timestamp too old');
  }

  // Compute expected signature
  const payloadString = typeof payload === 'string' ? payload : payload.toString('utf8');
  const signedPayload = `${timestamp}.${payloadString}`;
  const expectedSignature = crypto
    .createHmac('sha256', secret)
    .update(signedPayload)
    .digest('hex');

  // Compare signatures (constant time to prevent timing attacks)
  if (!crypto.timingSafeEqual(Buffer.from(expectedSignature), Buffer.from(providedSignature))) {
    throw new WebhookVerificationError('Invalid signature');
  }

  // Parse and return event
  try {
    return JSON.parse(payloadString);
  } catch (error) {
    throw new WebhookVerificationError('Invalid JSON payload');
  }
}

/**
 * Constructs webhook signature (for testing purposes)
 * @param payload Webhook payload
 * @param secret Webhook secret
 * @returns Signature string
 */
export function constructWebhookSignature(payload: string | object, secret: string): string {
  const timestamp = Math.floor(Date.now() / 1000);
  const payloadString = typeof payload === 'string' ? payload : JSON.stringify(payload);
  const signedPayload = `${timestamp}.${payloadString}`;
  const signature = crypto.createHmac('sha256', secret).update(signedPayload).digest('hex');
  return `t=${timestamp},v1=${signature}`;
}

File: packages/python-sdk/z402/utils/webhook.py
"""Webhook signature verification utilities"""

import hashlib
import hmac
import json
import time
from typing import Any, Dict, Union

from z402.exceptions import WebhookVerificationError


def verify_webhook(
    payload: Union[str, bytes, Dict[str, Any]],
    signature: str,
    secret: str,
    tolerance: int = 300,
) -> Dict[str, Any]:
    """
    Verify webhook signature and return parsed event.

    Args:
        payload: Raw request body (string, bytes, or dict)
        signature: Signature from z402-signature header
        secret: Your webhook secret
        tolerance: Maximum age of signature in seconds (default: 300)

    Returns:
        Parsed webhook event

    Raises:
        WebhookVerificationError: If signature is invalid

    Example:
        ```python
        @app.post("/webhooks/z402")
        async def handle_webhook(request: Request):
            signature = request.headers.get("z402-signature")
            body = await request.body()

            try:
                event = verify_webhook(body, signature, webhook_secret)
                # Handle event
                return {"received": True}
            except WebhookVerificationError as e:
                return {"error": str(e)}, 400
        ```
    """
    if not signature:
        raise WebhookVerificationError("Missing signature header")

    if not secret:
        raise WebhookVerificationError("Webhook secret not provided")

    # Parse signature (format: t=timestamp,v1=signature)
    parts = signature.split(",")
    timestamp_part = next((p for p in parts if p.startswith("t=")), None)
    signature_part = next((p for p in parts if p.startswith("v1=")), None)

    if not timestamp_part or not signature_part:
        raise WebhookVerificationError("Invalid signature format")

    timestamp_str = timestamp_part.split("=", 1)[1]
    provided_signature = signature_part.split("=", 1)[1]

    # Check timestamp is within tolerance
    try:
        timestamp = int(timestamp_str)
    except ValueError:
        raise WebhookVerificationError("Invalid timestamp in signature")

    now = int(time.time())
    if abs(now - timestamp) > tolerance:
        raise WebhookVerificationError("Signature timestamp too old")

    # Convert payload to string
    if isinstance(payload, dict):
        payload_string = json.dumps(payload, separators=(",", ":"))
    elif isinstance(payload, bytes):
        payload_string = payload.decode("utf-8")
    else:
        payload_string = payload

    # Compute expected signature
    signed_payload = f"{timestamp}.{payload_string}"
    expected_signature = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    # Compare signatures (constant time to prevent timing attacks)
    if not hmac.compare_digest(expected_signature, provided_signature):
        raise WebhookVerificationError("Invalid signature")

    # Parse and return event
    try:
        if isinstance(payload, dict):
            return payload
        return json.loads(payload_string)
    except json.JSONDecodeError:
        raise WebhookVerificationError("Invalid JSON payload")


def construct_webhook_signature(
    payload: Union[str, Dict[str, Any]],
    secret: str,
) -> str:
    """
    Construct webhook signature for testing.

    Args:
        payload: Webhook payload (string or dict)
        secret: Webhook secret

    Returns:
        Signature string in format: t=timestamp,v1=signature

    Example:
        ```python
        payload = {"type": "payment.settled", "data": {...}}
        signature = construct_webhook_signature(payload, "secret")
        ```
    """
    timestamp = int(time.time())

    if isinstance(payload, dict):
        payload_string = json.dumps(payload, separators=(",", ":"))
    else:
        payload_string = payload

    signed_payload = f"{timestamp}.{payload_string}"
    signature = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return f"t={timestamp},v1={signature}"
